app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt

```diff
@@ -204,126 +173,57 @@ class ChainServiceServer(val storage: TrustChainStorage, val me: ChainService.Pe
-                ", validation result: " + validation!!.toString())
-
-        // only send signedBlock if validated correctly
-        // If you want to test the sending of blocks and don't care whether or not blocks are valid, remove the next check.
-        if (validation.getStatus() !== ValidationResult.ValidationStatus.PARTIAL_NEXT && validation.getStatus() !== ValidationResult.ValidationStatus.VALID) {
-            Log.e(TAG, "Signed signedBlock did not validate. Result: " + validation.toString() + ". Errors: "
-                    + validation.getErrors().toString())
-            return null
-        } else {
-            storage.insertInDB(signedBlock)
-            Log.i(TAG, "saving freshly created block")
-            return signedBlock
-        }
-    }
-
-    fun signBlock(peer: ChainService.Peer, linkedBlock: MessageProto.TrustChainBlock): MessageProto.TrustChainBlock? {
-        // do nothing if linked block is not addressed to me
-        val myPublicKey = me.publicKey.toByteArray()
-        if (!Arrays.equals(linkedBlock.linkPublicKey.toByteArray(), myPublicKey)) {
-            Log.e(TAG, "signBlock: Linked block not addressed to me.")
-            return null
-        }
-        // do nothing if block is not a request
-        if (linkedBlock.linkSequenceNumber != TrustChainBlock.UNKNOWN_SEQ) {
-            Log.e(TAG, "signBlock: Block is not a request.")
-            return null
-        }
-        var block = TrustChainBlock.createBlock(null, storage,
-                myPublicKey,
-                linkedBlock, peer.publicKey.toByteArray())
-
-        block = TrustChainBlock.sign(block, keyPair.private)
-
-        val validation: ValidationResult?
-        try {
-            validation = TrustChainBlock.validate(block, storage)
-        } catch (e: Exception) {
-            e.printStackTrace()
-            return null
-        }
-
-        Log.i(TAG, "Signed block to " + Peer.bytesToHex(block.linkPublicKey.toByteArray()) +
-                ", validation result: " + validation!!.toString())
-
-        // only send block if validated correctly
-        // If you want to test the sending of blocks and don't care whether or not blocks are valid, remove the next check.
-        if (validation.getStatus() !== ValidationResult.ValidationStatus.PARTIAL_NEXT && validation.getStatus() !== ValidationResult.ValidationStatus.VALID) {
-            Log.e(TAG, "Signed block did not validate. Result: " + validation.toString() + ". Errors: "
-                    + validation.getErrors().toString())
-        } else {
-            storage.insertInDB(block)
-            return block
+        val peerChannel = registry.findStub(peerMessage)
+        val keySingle: Single<ChainService.Key> = peerChannel.sendLatestBlocks(crawlResponse).guavaAsSingle(Schedulers.computation())
+
+        return keySingle.flatMap { theirPublicKey ->
+            val newBlock = signerValidator.createNewBlock(payload, theirPublicKey.publicKey.toByteArray())
```

Comment:
`theirPublicKey.publicKey`: is theirPublicKey a Keypair by chance?

Reply:
no just the publicKey as a ChainService.Key (proto stuff) that we get back from the message. else that would mean we send the private key aswell.